Quality of life changes

lzardy opened the pull request:
Proper Minions class for use with skills and items. Change previous Minions to LaneMinions for later reworking and due to too many conflictions with trying to make previous Placeables class either inherit from Minions or being within the same class. Make Monsters class inherit from new Minions class. Change previous Minions packet to LaneMinions packet, fix Placeables packet, and change to Minions packet. Make use of the interfaces, cause why not and add INetworkIdManager as another interface. Add WithinRange function for spell scripting. For future fixing of server->client updates of projectile positions, add a comment for debugging. Add game reference for skills that spawn minions. Add EventOnHitUnit for additional scripting capabilities.

lzardy committed:
Add RefGame and rename IMinion

lzardy committed:
Another rename

lzardy committed:
Renames and more

Use interfaces, add AI targetting for new Minions, add RefGame for scripting, begin transition of uses of previous Minion to LaneMinion, Monsters inherit from new Minions, add EventOnHitUnit for scripting.

lzardy committed:
New Minion is part of AI

lzardy committed:
Finish renames, conversion ...

Rest of the interfaces, add WithinRange function for scripts, use proper NotifyMinionSpawned packet and add LeaguePackets.dll to references.

lzardy commented:
LeaguePackets.dll was not added to commit for some reason, adding.

lzardy committed:
Actually add LeaguePackets.dll

Not sure if anyone dislikes having packages folder before build, but this was the only way I could get the auto-reference to work.

cajthamlmatej commented:
~~Why did you removed Placeable? It's a thing from LoL (trinket, etc.) and you just can't change packets for it.~~ Lane minion should inherit from Minion (currently inherit from AiObjBase)

lzardy commented:
> Why did you removed Placeable? It's a thing from LoL (trinket, etc.) and you just can't change packets for it. Lane minion should inherit from Minion (currently inherit from AiObjBase)

I see no reason to inherit from new Minions besides naming conventions. What do you think it could benefit from inheriting?

cajthamlmatej commented:
Everything in lol is minion, lane minion is type of minion.

lzardy commented:
> Everything in lol is minion, lane minion is type of minion.

That doesn't answer my question.

yasinallahdev commented:
> > Everything in lol is minion, lane minion is type of minion.
> 
> That doesn't answer my question.

To my understanding, in League of Legend's source code, basically all of the unit types (Lane Minions, Monsters, Wards, the solid parts of Jarvan IV's wall) are all internally minions, with specific subtypes for Minions and Monsters. Using the new Minion class is essentially done for consistency reasons, and If I Recall Correctly, LaneMinion inherits from Minion in the official source as well.

lzardy commented:
> To my understanding, in League of Legend's source code, basically all of the unit types (Lane Minions, Monsters, Wards, the solid parts of Jarvan IV's wall) are all internally minions, with specific subtypes for Minions and Monsters. Using the new Minion class is essentially done for consistency reasons, and If I Recall Correctly, LaneMinion inherits from Minion in the official source as well.

The main problem I have is with new Minions, they have an owner, I haven't the slightest idea how to let lane minions be their own owner.
EDIT: Nevermind, it is done. EDIT 2: Lane Minions seem to be quite broken.

lzardy committed:
LaneMinions inherits from Minions

Possibly a dirty fix, but it works as though it still inherits from IObjAiBase.

lzardy commented:
PR is now broken due to trying to make LaneMinions inherit Minions, will update when it is fixed.

lzardy committed:
Fix issues with LeaguePackets

lzardy committed:
Actually upload LeaguePackets.dll (2nd time)

lzardy committed:
Fix path of LeaguePackets.dll

lzardy committed:
Merge branch 'simple-changes' of https://github.com/microsoftv/GameServer into simple-changes

lzardy committed:
Fix lane minions

lzardy commented:
LaneMinions inherits Minions, and LaneMinions act like normal. As an unexpected addition (good/bad?), !spawn command now spawns minions that attack and move to nearby targets. The PR is not broken anymore.

HarmfulBreeze commented on GameServerLib/GameObjects/AttackableUnits/AttackableUnit.cs:
```diff
@@ -30,15 +31,15 @@ public class AttackableUnit : GameObject, IAttackableUnit
 
         public AttackableUnit(
-            Game game,
+            IGame game,
```
Why not just keep the implementation? I don't understand your change, and it adds an unnecessary cast.
```suggestion
            Game game,
```

lzardy replied:
Used by scripts that spawn minions. Spawning requires game instance.

danil179 replied:
Where scripts use AttackableUnit? they should use only IAttackableUnit or similar things.

lzardy replied:
It uses IMinion: `IMinion Jack = new Minion(owner.RefGame, owner, spell.X, spell.Y, "ShacoBox", "ShacoBox", sightrange, 0);
                owner.RefGame.ObjectManager.AddObject(Jack);
                Jack.SetVisibleByTeam(owner.Team, true);`

danil179 replied:
I'm not sure this is in the cur build, that shouldn't be the case, the API should wrap this.

lzardy replied:
In my PR, it uses IMinion.

danil179 replied:
ok, so we should open issue about API wrapping. There should be API call like `AddMinion(...)` etc.

lzardy replied:
I agree, but would it come before or after my PR?

danil179 replied:
After, but I don't see the minion creation code here: https://github.com/LeagueSandbox/LeagueSandbox-Default/pulls, so maybe you didn't PRed the submodule.

It's important to know that changes to the submodule don't appear when you PR to GameServer. Only when you PR the submodule and add change the submodule pointer it will appear.

danil179 replied:
Fix here as suggested, It should be possible after introducing the API `AddMinion` function.

HarmfulBreeze commented on GameServerLib/GameObjects/AttackableUnits/AttackableUnit.cs:
```diff
@@ -30,15 +31,15 @@ public class AttackableUnit : GameObject, IAttackableUnit
             int visionRadius = 0,
             uint netId = 0
-        ) : base(game, x, y, collisionRadius, visionRadius, netId)
+        ) : base((Game)game, x, y, collisionRadius, visionRadius, netId)
```
Goes with the one above.
```suggestion
        ) : base(game, x, y, collisionRadius, visionRadius, netId)
```

danil179 replied:
Also fix here as suggested.

HarmfulBreeze commented on GameServerLib/Packets/INetworkIdManager.cs:
```diff
@@ -0,0 +1,7 @@
+﻿namespace GameServerCore
+{
+    public interface INetworkIdManager
```
Why have an interface for our NetID manager? The interface will only ever be implemented by the actual manager, so there's no point in having it.
Please remove the interface. (and revert your other changes related to it)

danil179 replied:
I think that scripts should never use NetID and other lib related stuff, so we don't need interface for that and should manage things behind the scenes.

HarmfulBreeze commented on PacketDefinitions420/PacketNotifier.cs:
```diff
@@ -87,7 +88,7 @@ public void NotifyPing(ClientInfo client, Vector2 pos, int targetNetId, Pings ty
         }
 
-        public void NotifyTint(TeamId team, bool enable, float speed, Color color)
+        public void NotifyTint(TeamId team, bool enable, float speed, GameServerCore.Content.Color color)
```
Add a ``using`` directive.

lzardy replied:
Already has one, I am using LeaguePackets and GameServerCore.Content, both Color, in them so I have to specify which one.

MythicManiac replied:
I find this ambiguity itself to be an issue. It's a separate issue, but maybe the two should either be unified or one of them renamed.

lzardy replied:
I feel it may be easier in general if we just added LeaguePackets project and edit from there.

microsoftv replied:
I feel it may be easier in general if we just added LeaguePackets project and edit from there.

HarmfulBreeze commented on PacketDefinitions420/PacketNotifier.cs:
```diff
@@ -273,7 +274,7 @@ public void NotifyHeroSpawn(int userId, ClientInfo client, int playerId)
         public void NotifyAvatarInfo(int userId, ClientInfo client)
         {
-            var info = new AvatarInfo(client);
+            var info = new PacketDefinitions.S2C.AvatarInfo(client);
```
Add a ``using`` directive.

lzardy replied:
Same as above.

HarmfulBreeze commented on PacketDefinitions420/PacketNotifier.cs:
```diff
@@ -401,7 +402,7 @@ public void NotifyMovement(IGameObject o)
         }
 
-        public void NotifyDamageDone(IAttackableUnit source, IAttackableUnit target, float amount, DamageType type, DamageText damagetext, bool isGlobal = true, int sourceId = 0, int targetId = 0)
+        public void NotifyDamageDone(IAttackableUnit source, IAttackableUnit target, float amount, GameServerCore.Enums.DamageType type, DamageText damagetext, bool isGlobal = true, int sourceId = 0, int targetId = 0)
```
Add a ``using`` directive.

lzardy replied:
Same as above.

HarmfulBreeze reviewed (changes requested):
Needs a few changes.
Overall a pretty good PR, we need more PRs like this one in the future (splat in multiple PRs if possible)

P.S.: why do we have ``IGame`` when it's only ever implemented by Game?

danil179 commented on PacketDefinitions420/PacketDefinitions/S2C/SpawnMinion.cs:
```diff
@@ -4,33 +4,33 @@
+    public class SpawnMinion : Packet
     {
-        public SpawnPlaceable(IPlaceable p)
+        public SpawnMinion(IMinion m)
```
I think you can just delete this file and use LeaguePackets version. this version of 4.20 is not actually correct, and not used (after spawn placeable change).

danil179 commented on GameServerCore/Domain/GameObjects/ILaneMinion.cs:
```diff
@@ -0,0 +1,10 @@
+{
+    public interface ILaneMinion : IObjAiBase
+    {
+        MinionSpawnPosition SpawnPosition { get; }
```
see issue #783 if you can fix that within this PR.

danil179 commented on GameServerCore/Domain/GameObjects/ITarget.cs:
```diff
@@ -10,5 +10,6 @@ public interface ITarget
 
         float GetDistanceTo(ITarget target);
         Vector2 GetPosition();
+        bool WithinRange(Vector2 from, Vector2 to, float range);
```
If is this a static method why is it in ITarget? or just use ITarget position.

lzardy replied:
It is used for abilities that can be cast at any range but that shouldn’t be, like Shaco W.

danil179 replied:
still, this function is sort of helper function. Either use some variables from ITarget/classes implement this function, or make this function static.

lzardy replied:
Making it static will not allow me to use: `owner.WithinRange()` for spells, since IChampion doesn't have WithinRange().

danil179 replied:
you will use `HelperLib.WithinRange(...)`

wildbook replied:
I would suggest keeping WithinRange as a method on (in the Vector2 class), or extension method for (HelperLib approach), Vector2. Mainly considering that it's most likely going to be used for a lot more than just targeting in the future.

That way you could do something like `owner.Location.IsWithinRange(target.Location, range)`. You could also add an overload taking ITarget instead of Vector2 and that way allow `owner.Location.IsWithinRange(target, range)`

danil179 replied:
Maybe ITarget can just inherit Vector2 no? Then it will work with ITarget as well.

Im voting for the extension functions (of Vector2) solution.

wildbook replied:
As far as I'm aware, an interface can't inherit from a class. I had the same idea starting out but scrapped it due to that. You could however make an IVector2 interface and use that if you wanted to, then inherit from that in both Vector2 and ITarget.

I'm also most likely for the extension method approach, alternatively a method on Vector2 and an extension function that takes and converts ITarget to Vector2 then passes that to the method on Vector2, that way keeping Vector2 cleaner.

lzardy replied:
I use WithinRange for if spell.X and spell.Y are within correct cast range.

danil179 replied:
The point is that this is `static` function when you want to use non-static in ITarget. That creates some ambiguity so we need this as extended or helper function. (some changes in hierarchy needed)

danil179 replied:
For another PR.

danil179 commented on GameServerLib/API/ApiEventManager.cs:
```diff
@@ -152,4 +153,36 @@ public void Publish(IChampion champion)
+        {
+            listeners.ForEach((listener) =>
+            {
+                if (listener.Item2 == unit)
```
If you really do reference checking you need to use just HashTable or something similar and not List.

lzardy replied:
I have no experience with HashTables, any ideas?

danil179 replied:
so Ill open issue about this. Just use C# Dictionary type, there are many.

danil179 replied:
Opened issue #796 .

danil179 commented on GameServerLib/GameObjects/AttackableUnits/AI/LaneMinion.cs:
```diff
@@ -0,0 +1,182 @@
+            MinionSpawnPosition position,
+            List<Vector2> mainWaypoints,
+            uint netId = 0
+        ) : base(game, null, 0, 0, "", "", 1100, netId)
```
Do you really want to pass null, why?

lzardy replied:
I didn't know any alternatives, perhaps it should be Nexus or barracks, is that right?

danil179 replied:
`null` is bad usually, what you want is to pass some default object, it depends on the context.

danil179 replied:
`null` here stands for the owner. The owner should be the object itself probably or the barracks created it, depends on packets. Currently I think null is ok, but we should change Minion class to accept no owner.

danil179 commented on GameServerLib/GameObjects/AttackableUnits/AI/Minion.cs:
```diff
@@ -1,73 +1,42 @@
+            int visionRadius = 0,
             uint netId = 0
-        ) : base(game, "", new Stats.Stats(), 40, 0, 0, 1100, netId)
+        ) : base((Game)game, model, new Stats.Stats(), 40, x, y, visionRadius, netId)
```
If you use an actual class and not the interface (Minion and not IMinion) then there is no problem to use the actual Game class.

If there is where is it?

danil179 commented on GameServerLib/GameObjects/AttackableUnits/AI/Minion.cs:
```diff
@@ -78,7 +47,10 @@ public void PauseAi(bool b)
         {
             base.OnAdded();
-            _game.PacketNotifier.NotifyMinionSpawned(this, Team);
+            if (!(Owner == null))
```
Why no owner is a problem? if it is because of how Notify implemented then put the null check there.

danil179 replied:
Packet change if no owner so we need to check, resolved.

danil179 commented on GameServerLib/GameObjects/Monster.cs:
```diff
@@ -29,7 +29,7 @@ public class Monster : ObjAiBase, IMonster
             float spawnAnimationTime = 0.0f,
             uint netId = 0
-        ) : base(game, model, new Stats.Stats(), 40, x, y, 0, netId)
+        ) : base(game, null, x, y, model, name, 0, netId)
```
again, explain why you use null.

danil179 replied:
I consider this resolved as in `LaneMinion` class cause no owner exist.

danil179 commented on GameServerLib/GameObjects/Other/Target.cs:
```diff
@@ -47,5 +47,11 @@ public Vector2 GetPosition()
             return new Vector2(X, Y);
         }
+
+        public bool WithinRange(Vector2 from, Vector2 to, float range)
```
Really should be static and be putted in some other helper class.

Or to use the ITarget position somehow.

lzardy replied:
Would this be better in AttackableUnit?

danil179 replied:
try to do that as extended function like:
`to.WithinRange(from,range)`

And put the extended function in some helper class in core. There should be already some helper functions.

danil179 replied:
For another PR.

danil179 reviewed (changes requested):
Reviewed.

I see that mainly you changed IPlaceable to IMinion. Many things can be cleaned up. I think that you should refactor some more things before merge but these not really you fault. Some of your added interfaces in my POV are unnecessary.

lzardy committed:
Add Minion Replication

lzardy committed:
Show pet DmgText and IsPet/IsWard

If the minion is a pet, then damage text that the minion has shows for the owner. IsPet is used for abilities that spawn minions. IsWard is not implemented yet, but it would likely be used for abilities that spawn wards.

lzardy committed:
Small fixes

Use correct owner in Minion.cs and change NotifyMinionSpawned to not always set IsWard to true.

lzardy committed:
Resolve unused if

lzardy committed:
AddMinion() for scripting

lzardy committed:
Remove RefGame and INetworkManager

lzardy committed:
SpawnMinion incorrect and replaced by LeaguePackets

lzardy committed:
Delete SpawnMinion.cs from .csproj

lzardy committed:
Fix LaneMinions

Spawning LaneMinions with NotifyMinionSpawned causes them to stand still at their spawn position and teleport to targets.

danil179 commented:
Please open issues on both projects that explain the problems solved or features added by your PR.

danil179 commented on GameServerLib/API/ApiFunctionManager.cs:
```diff
@@ -149,6 +149,30 @@ public static void RemoveParticle(Particle p)
+            var m = new Minion(_game, champion, target.X, target.Y, model, name, visionRadius);
+            _game.ObjectManager.AddObject(m);
+            m.SetVisibleByTeam(champion.Team, true);
+            _game.PacketNotifier.NotifyMinionSpawned(m, champion.Team);
```
`AddObject` already send `NotifyMinionSpawned`. (see `OnAdded` function for minion)

danil179 commented on GameServerLib/API/ApiFunctionManager.cs:
```diff
@@ -149,6 +149,30 @@ public static void RemoveParticle(Particle p)
+        public static void RemoveMinion(IMinion m, IAttackableUnit killer)
+        {
+            _game.ObjectManager.RemoveObject(m);
+            _game.PacketNotifier.NotifyNpcDie(m, killer);
```
Consider moving this to `OnRemoved` function for `Minion`.

danil179 replied:
Currently don't move Die to `OnRemoved` (cause it will probably will cause some breaks in the game mechanics), Ill open issue about this to fix the whole remove system later.

danil179 commented on GameServerLib/GameObjects/AttackableUnits/AI/Minion.cs:
```diff
@@ -57,7 +58,10 @@ public void PauseAi(bool b)
         {
             base.OnAdded();
-            _game.PacketNotifier.NotifyMinionSpawned(this, Team);
+            if (!(Owner == null))
```
Consider moving this check into `NotifyMinionSpawned`.

danil179 replied:
Actually I see that you also check for `null` Owner in `NotifyMinionSpawned`. So is it that always `Minion` needs an owner? why? please explain about this cause I find this concept confusing.

lzardy replied:
To quote the commit "Fix LaneMinions":
> Spawning LaneMinions with NotifyMinionSpawned causes them to stand still at their spawn position and teleport to targets (basically they don't act normal).

If there was no check for null owner in Minion.cs, this issue would occur. I think this is because NotifyMinionSpawned still has issues with updating minion position after spawn. If I make LaneMinions only spawn with NotifyLaneMinionSpawned, this issue does not occur, which is what the null check is for.

danil179 replied:
Just disable in LaneMinion the call to `base.OnAdded();` (but check if that doesn't break anything - like things in IObjBase.

danil179 replied:
Resolved here.

danil179 reviewed (commented):
Added more comments.

lzardy committed:
Clean and Organize

NotifyMinionSpawned variables assigned according to SpawnMinionS2C() order.
Added IsBot, IsLaneMinion, and IsClone according to SpawnMinionS2C() requirements for later changes. Only Minion.cs has check for null Owner.

danil179 commented on GameServerLib/GameObjects/AttackableUnits/AI/Minion.cs:
```diff
@@ -25,18 +28,34 @@ public class Minion : ObjAiBase, IMinion
-            Owner = owner;
+            if (this is ILaneMinion)
+            {
+                IsLaneMinion = true;
```
Consider put this in `LaneMinion` constructor.

lzardy replied:
I am going to delete LaneMinion altogether, that's fine right?

danil179 replied:
Why? We at the end accepted that it should be inherited. Just change how you use the base class (Don't send `base.OnAdded` and such things.

lzardy replied:
> yeah, probably delete this class/interface. Just bool variable in minion class.

Is what you said in response to:

> I think the idea of having LaneMinion be seperate from Minion is not accurate to league anyway.

Am I misinterpreting it? We already inherit from Minions for LaneMinions.

danil179 replied:
At the end we concluded that you should remove this bool and just use Minion class in the right way from `LaneMinion` class. Like sending the 3 packet instead of Minion packet.
The problem was in the packets (we use wrong packets)

danil179 replied:
And as we concluded later, also `IsLaneMinion` flag is ok. (what you did)

lzardy replied:
> And as we concluded later, also `IsLaneMinion` flag is ok. (what you did)

Now it is just which option is more clean, as currently we have LaneMinions and Minions as seperate classes, with LaneMinion inheriting from Minions. With the other option; them combined, functions that used to use LaneMinion(), would be changed to Minion() with Minion() now having MinionSpawnType, MinionSpawnPosition, and mainWaypoints as optional parameters that can be put into Minion(**here**). Since they are optional parameters, any current uses of Minion() would stay the same, but all uses of LaneMinion() would have to be changed to account for the combination of the two functions:

`new Minion(Game, champion, 0, 0, "", "", 1100, 0, MinionSpawnType.MINION_TYPE_SUPER, spawnPositions[team])` 
with these as the parameters:
`Game game, IAttackableUnit owner, float x, float y, string model, string name, int visionRadius = 0, uint netId = 0, MinionSpawnType spawnType = 0x00, MinionSpawnPosition position = 0x00, List<Vector2> _mainWaypoints = null`
In which `int visionRadius = 0` is where the optional parameters start.

danil179 replied:
`IsLaneMinion` is league concept which we dont yet know how it is actually used. I consider this as solved but opened issue to understand what `IsLaneMinion` flag is for. Opened issue  #797

lzardy committed:
Merge pull request #1 from LeagueSandbox/indev

Re-base

lzardy committed:
Fix checks and begin re-base

lzardy committed:
Resolve Conflicts for re-base

lzardy committed:
Merge branch 'indev' into simple-changes

lzardy committed:
Resolve build errors from re-basing

lzardy committed:
Remove Leftovers

danil179 commented on GameServerLib/GameObjects/AttackableUnits/AI/LaneMinion.cs:
```diff
@@ -0,0 +1,181 @@
+        protected int _curMainWaypoint;
+        public MinionSpawnPosition SpawnPosition { get; }
+        public MinionSpawnType MinionSpawnType { get; }
+        protected bool _aiPaused;
```
Already in `Minion` class, remove this.

danil179 commented on GameServerLib/GameObjects/AttackableUnits/AI/LaneMinion.cs:
```diff
@@ -0,0 +1,181 @@
+
+        }
+
+        public void PauseAi(bool b)
```
Remove this, already in `Minion` class (should be the same).

danil179 commented on GameServerLib/API/ApiFunctionManager.cs:
```diff
@@ -149,6 +149,28 @@ public static void RemoveParticle(Particle p)
+            return m;
+        }
+
+        public static void RemoveMinion(IMinion m, IAttackableUnit killer)
```
Add to this function the calling to `m.Die(killer)`  instead of the `_game.PacketNotifier.NotifyNpcDie(m, killer);` to handle some additional things (gold, xp etc).

I think that `Die` function works, cause it actually call `SetToRemove()` (see `AttackableUnit`) which marking this object to removal in the next `Update` tick of the `ObjectManager`.

First try not using `RemoveMinion` and only `Die`, it should work.

lzardy replied:
Using m.Die() causes turrets to forever target the unit if the turret was **about** to shoot it (when there is a spark on the turret, but no projectile). I think this is a combination of turret targeting issues and Die() function not being entirely functional.

danil179 reviewed (commented):
Some additional comments.

lzardy committed:
Remove LaneMinion PauseAi

lzardy committed:
RemoveMinion removed

Was hiding a bigger issue regarding targeting, Die() function should be fine enough to use as a replacement, however turrets seem to forever target dead minions when the turret is about to shoot and the minion dies before the projectile appears.

danil179 commented:
Before merging, please open issues on both projects that explain the problems solved or features added by your PR. (for example, API lacks function for adding minions or Placeable shouldn't exist, etc)
